doc/threshold/SONiC Threshold feature spec.md

```diff
@@ -0,0 +1,589 @@
+
+2.0.6.5 The breach entry must contain the buffer usage in terms of percentage at the time of breach.
+
+2.0.6. The breach entry must contain the absolute buffer usage in bytes at the time of breach.
+
+2.0.6.4 The breach entry must contain the time stamp of the breach event. The ASIC provides the time stamp of breach event via the breach event protobuf.
+
+2.0.7 The threshold feature must be a part of the TAM container along with other TAM features.
+
+3.0 The threshold feature must be supported on physical ports. LAG member ports will continue to be tracked at the physical port level.
+
+3.1 The threshold feature does not support dynamic port breakout.
+
+3.2 In case of static port breakout, the user is expected to reconfigure the broken out ports.
+
+3.3 CPU port counters for queues are not supported for threshold configuration.
+
+4.0 The threshold can be configured in percentage mode only.
+
+4.1 The breach entry in COUNTERS\_DB must show the usage at breach in bytes and percentage.
+
+5.0 UI commands available to check currently configured thresholds.
+
+5.1 UI commands available to show threshold breach events.
+
+
+### 1.1.2 Configuration and Management Requirements
+
+The threshold feature will use the python click framework for CLI. 
+
+### 1.1.3 Scalability Requirements
+There is no limit enforced on the threshold configuration and breach event entries.
+
+## 1.2 Design Overview
+### 1.2.1 Basic Approach
+The threshold feature is newly developed.  
+
+### 1.2.2 Container
+A new container, "tam" is created to host the ThresholdMgr. 
+
+### 1.2.3 SAI Overview
+The SAI TAM spec specifies the TAM APIs to be used to configure the threshold functionality. Please refer section 3.5 for more details.
+
+# 2 Functionality
+## 2.1 Target Deployment Use Cases
+
+The threshold feature can be used as a monitoring tool to detect congestion in the network. The breach information is available in COUNTERS\_DB thereby allowing applications to be built upon data from the threshold and watermark features to manage and monitor networks.
+
+## 2.2 Functional Description
+
+In a live network, tracking and monitoring buffer usage on a switch can be extremely useful in detecting and isolating potential and actual congestion events in the network. The availability of telemetry information w.r.t buffer usage on network nodes can help to develop network-level monitoring applications. 
+
+The watermark feature in SONiC makes the buffer usage counters readily available in COUNTERS\_DB for multiple users. The watermarks can be used to track peak buffer occupancy across users and intervals. In addition to this, it may also be useful for the user to know when the buffer usage crosses a certain threshold. The threshold feature implements this specific functionality.
```

Comment:
Can the spec support CPU queues thresholds?

Reply:
As of now, the threshold feature supports the same buffers as watermark feature i.e front panel port's priority-groups and queues. However, the idea is to support buffer-pool, CPU queues as well for threshold feature going forward.